### PR TITLE
RFC: More common updater script implementations

### PR DESCRIPTION
**Pull request by dezgeg**
As a first step of adding some common updater script implementations, this PR adds a script to automate patching the `version` and `sha256` fields in .nix files. It currently uses the hacky way of replacing the existing hash with an invalid one and parsing the proper hash from the error message. When https://github.com/NixOS/nix/issues/1172 is fixed, `nix-build --hash` can be used.

Now, others have proposed (and used) different and arguably cleaner approaches than patching the .nix files, like generating separate JSON files for sources (e.g. https://github.com/NixOS/nixpkgs/pull/21734). But from a pragmatic point of view, we already have thousands of packages using the traditional "version/sha in .nix file" style, so it likely makes sense to make the tooling work with those instead of trying to convert all of them.

cc @garbas for the `tp_smapi` stuff and @taku0 for the Firefox/Thunderbird stuff. Also @edolstra for the general approach & locations of the scripts.

For easier testing, here are the same commits + the relevant packages downgraded: https://github.com/dezgeg/nixpkgs/commits/updater-downgraded

### CODE REVIEW

**Comment by taku0** [comment]
I prefer generating separate JSON file. I should have done so for Firefox/Thunderbird. After all, we need rewrite each packages to adopt `updateScript`, which may not a simple task. We may need to write a web scraper to know the latest version. Compared to it, rewriting `default.nix` to using separate JSON file is a simple task.

**Comment by taku0** [comment]
Of course, this update script is far better than my ad hoc update script using `ed`.

**Comment by dezgeg** [comment]
I think in a large number of cases rewriting the packages is not necessary and things like figuring out the latest version can be automated. E.g. if a package is currently using `fetchgit` to fetch a particular tag, it's very likely that an automated upgrade could just update to a newer tag in the repo. Similarly, for packages using tarballs all that might be needed is to specify `meta.downloadPage` and see if there are links to newer tarballs following the same version pattern.

(This is the similar logic what http://monitor.nixos.org/ used to do when it was up and I found it very useful).

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix [hunk 1]
```diff
@@ -130,11 +129,8 @@ stdenv.mkDerivation rec {
     platforms = platforms.linux;
   };
 
-  passthru.updateScript = import ./../../browsers/firefox/update.nix {
-    name = "thunderbird";
-    sourceSectionRegex = ".";
-    basePath = "pkgs/applications/networking/mailreaders/thunderbird";
+  passthru.updateScript = callPackage ./../../browsers/firefox/update.nix {
```
Comment (edolstra) [inline pkgs/applications/networking/mailreaders/thunderbird/default.nix]: This is an abuse of `callPackage`. `update.nix` is not a package...

Comment (dezgeg) [inline pkgs/applications/networking/mailreaders/thunderbird/default.nix]: Should we have the same thing under a different name? It's quite painful to duplicate the list of upgrade script dependencies four times.

File: maintainers/scripts/common-updater/update-source-version [hunk 2]
```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env nix-shell
+#! nix-shell --pure -p bash nix coreutils gawk gnused -i bash
+set -e
+
+# FIXME: It's dumb that this is needed.
```
Comment (edolstra) [inline maintainers/scripts/common-updater/update-source-version]: Why is this needed?

Comment (dezgeg) [inline maintainers/scripts/common-updater/update-source-version]: Otherwise this happens (on NixOS):
````
error: Nix database directory ‘/nix/var/nix/db’ is not writable: Permission denied
./maintainers/scripts/common-updater/update-source-version: error: Couldn't evaluate 'thunderbird.meta.position' to locate the .nix file!
````

**Review by garbas — CHANGES_REQUESTED** [review]
i'm not a fan of using nix-shell inside nix-shell. i would rather have this common-scripts become packages in nixpkgs (under some namespace) and which we run them using ``maintainers/scripts/update.nix``

**Comment by edolstra** [comment]
This looks basically good to me. Having a 90-line shell script full of grep/sed magic is not ideal but that could be improved in the future. (I'm thinking that ideally, we would have `nix edit-attr` command that would use the evaluator knowledge of position info in the AST. That could also be useful for editing NixOS configurations programmatically...)

**Comment by edolstra** [comment]
@garbas I agree.

@dezgeg Regarding `callPackage`, I think the real issue is abstracting away boilerplate code in update scripts. We really don't want to pass in `coreutils` etc. in every update script. See my comment here: https://github.com/NixOS/nixpkgs/pull/21405/files#r95163181. Such a `mkUpdateScript` function would presumably also have `update-source-version` in its default `PATH`.

**Comment by dezgeg** [comment]
Okay, thank you for the feedback. I've now moved this to `pkgs/common-updater/scripts.nix`, with the intention of placing the `mkUpdateScript` function Eelco proposed into `pkgs/common-updater/default.nix` later on. I have some ideas (and unfinished code) on the overall structure and things like common implementation for updating a `fetchFromGitHub` package to the latest tag in the repository that I will try to polish up soon-ish.

**Comment by dezgeg** [comment]
Any issues with the updated version?

**Comment by dezgeg** [comment]
I have pushed this as it also fixes the `updateScript` for `tp_smapi` that got broken by c20cc6d0b37b8330eb1bcdc0031b600f2665eb93.
